spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/MeanAbsolutePercentageCriterion.scala

```diff
@@ -0,0 +1,130 @@
+ * and return 100 * K.mean(diff) as outpout
+ * Here, the x and y can have or not have a batch.
+ * @param ev$1
+ * @param ev
+ * @tparam T The numeric type in the criterion, usually which are [[Float]] or [[Double]]
+ */
+class MeanAbsolutePercentageCriterion[T: ClassTag]
+  (implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
+
+  @transient
+  private var buffer1 : Tensor[T] = null
+  @transient
+  private var buffer2 : Tensor[T] = null
+  private val epsilon: T = ev.fromType(1e-07)
+  private val maxValue: T = ev.fromType(Double.MaxValue)
+  private val negtiveOne: T = ev.fromType(-1)
+
+
+  override def updateOutput(input: Tensor[T], target : Tensor[T]): T = {
+    if (buffer1 == null) buffer1 = Tensor[T]()
+    if (buffer2 == null) buffer2 = Tensor[T]()
+    buffer1.resizeAs(input).copy(input)
+    buffer2.resizeAs(input).copy(input)
+    buffer1.sub(target).abs()
+    buffer2.apply1(e => ev.clip(ev.abs(e), epsilon, maxValue))
+    buffer1.div(buffer2)
+
+    output = ev.times(buffer1.mean(), ev.fromType(100.0))
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], target: Tensor[T]): Tensor[T] = {
+    val norm : T = ev.fromType(100.0 / input.nElement())
+
+    buffer1.resizeAs(input).copy(input)
+    buffer2.resizeAs(target).copy(target)
+    gradInput.resizeAs(input).copy(input)
+
+    val func = new TensorFunc6[T] {
+      override def apply(data1: Array[T], offset1: Int, data2: Array[T], offset2: Int,
+                         data3: Array[T], offset3: Int): Unit = {
+        val a = data1(offset1)
+        val b = data2(offset2)
+
+        if (ev.isGreater(a, b)) { // x > y
+          if (ev.isGreaterEq(ev.abs(a), epsilon) && ev.isGreaterEq(maxValue, ev.abs(a))) {
+            if (ev.isGreater(a, ev.zero)) {
+              data3(offset3) = ev.divide(b, ev.times(a, a)) // y/x^2
+            } else {
+              data3(offset3) = ev.divide(b, ev.times(a, a))
+              data3(offset3) = ev.times(data3(offset3), negtiveOne) // -y/x^2
+            }
+          } else if (ev.isGreater(epsilon, ev.abs(a))) {
+            data3(offset3) = ev.divide(ev.one, epsilon) // 1/epsilon
+          } else {
+            data3(offset3) = ev.divide(ev.one, maxValue) // 1/Double.MaxValue
+          }
+        } else if (ev.isGreater(b, a)) { // x < y
+          if (ev.isGreaterEq(ev.abs(a), epsilon) && ev.isGreaterEq(maxValue, ev.abs(a))) {
+            if (ev.isGreater(a, ev.zero)) {
+              data3(offset3) = ev.divide(b, ev.times(a, a))
+              data3(offset3) = ev.times(data3(offset3), negtiveOne) // -y/x^2
+            } else {
+              data3(offset3) = ev.divide(b, ev.times(a, a)) // y/x^2
+            }
+          } else if (ev.isGreater(epsilon, ev.abs(a))) {
+            data3(offset3) = ev.divide(negtiveOne, epsilon) // -1/epsilon
+          } else {
+            data3(offset3) = ev.divide(negtiveOne, maxValue) // -1/Double.MaxValue
+          }
+        } else {
```

Comment:
If the only difference between x >y and x<y is the sign, maybe we can avoid the code duplicate here.

```
if ( x = y) {

} else {
   // calculate data3
   if (x < y) {
       data3 *= -1
    }
}
```

Reply:
ok, good idea